### ISSUE TITLE
Child links don't show when published if items are in the same document

### ISSUE THREAD

**Issue by samvrlewis** [doorstop-dev/doorstop#445]
Hi and thanks for the great tool!

We're trialling using doorstop and came across this strange behaviour when linking items within a single document. 

It appears that linking between items in a single document works, but only the parent link is shown when publishing a report on the requirements.

For example, using the following commands:

```
doorstop create SRD ./reqs
doorstop add SRD
doorstop add SRD
doorstop link srd001 srd002
doorstop publish srd
```

Produces the following output:

```
1.0     SRD001

        Parent links: SRD002

1.1     SRD002
```

Note that SRD002 does not have a 'Child links' list.

Looking through the source code, this appears to be because of a check in item.py:

https://github.com/doorstop-dev/doorstop/blob/5a074ad3918216e8cd956464119092b330612f87/doorstop/core/item.py#L730

Which, if I understand it correctly, means that only child links will be found if the child is in a child document to the parent's item. 

If I manually remove this check then everything works as I'd expect it to but it does beg a few questions:

- Why is that check there? Is it not valid to have links between requirements within the same document?
- If it is not valid to have links between requirements within the same document, should the tool not allow this to happen in the first place?

As it stands, this is quite confusing to me, so I'd appreciate any feedback. Happy to contribute some code to use my expected behaviour either as default, or through a configurable setting.

**Comment by jacebrowning** [comment]
> If it is not valid to have links between requirements within the same document, should the tool not allow this to happen in the first place?

I'm leaning toward a "yes" on this, but I'll let others weigh in.

**Comment by Spenser309** [comment]
I'd lean towards allowing this.  Imagine a single requirements document which breaks down higher level objectives.  

```
1.0 SRD001 
  Do X, Y and Z?

1.1 SRD002
  Do X

1.2 SRD003
 Do Y

1.3 SRD 004
 Do Z
```

Maybe X, Y, and Z have a single test but the implementation needs three different implementation tasks.  Or maybe X, Y, and Z are competing constraints and need to be tested together.

**Comment by sebhub** [comment]
I would not put too much policy into the tool and allow document internal links.

**Comment by jacebrowning** [comment]
Feel free to submit a patch with that check removed to see if it has any unintended consequences.

**Comment by samvrlewis** [comment]
> Feel free to submit a patch with that check removed to see if it has any unintended consequences.

Thanks, started working through a patch and realised (through running the tests) that making this change will break the concept of how traceability currently works.

The current 'row based' traceability works if links between requirements within the same document are not permitted, but maybe is not so suitable if this constraint is lifted. 

![image](https://user-images.githubusercontent.com/3880246/72024797-60616680-32ca-11ea-84fb-6bea522fc339.png)

What would others think of changing this to be more of a tree-like structure? Happy to try to knock something together though I fear that executing the visualisation of the tree might be the trickiest part. :)

**Comment by robbel** [comment]
@samvrlewis @jacebrowning could it not just be a bigger matrix where REQ, LLT, TUT, HLT (concatenated) make up both rows and columns? Then an "x" in any particular spot would mean that there is a link between that (row, col).

**Comment by sevendays** [comment]
IMHO (automotive):

**normative requirements must form a tree.**

The document tree doesn't fit the bill. Since plantuml can be used a tree should be easy to do:
![immagine](https://user-images.githubusercontent.com/301673/81833056-8714ff80-953f-11ea-94ed-13fdd63633c8.png)
[--> plantUML link](http://www.plantuml.com/plantuml/umla/ZL9DIyD04BtlhnZZHG-5slG3F4WFjVOaQ4IHsiEIdIPBazcuEqiFud-teNKR5W8zldrltnlUpDonjEphIhYTfaPQMKDjoEIbDJK2Muz2R4sbq4Aoz9IpDYIhBmABtrvRh97OWKDEu4TOp5bIKI4aw_NryldryMKulDji4f0E_g7uFJkkf5LxQV4YlzN7mI0Tf-EJOS0zkacacmWoZ61rKR9OO6xgnZXKuaQJmWP98UK8qXQ-7K8Gea9rAmv4-wydP1g5J7jqiqWtwzQGcIrVN4CMrH1mh8jg2BYR_8PvIJg_SAFPPzRkS3GwMGRSeqjFEWWFRNOO60P-yD0n2HF6H2ZtI9mN6jhi8oAhD13Zc0XfHo9SQr14xw9pJM2sdG-byPM2IjDEi04kiTii34JBW7QmDtQdgO2zvXAozoSmdXlFLt2_24StHrA-hlu0)

**parent requirements in the same document should be allowed** but a circular dependency should be forbidden. I don't get the need to have parents in a separate document.

**Comment by robbel** [comment]
@sevendays I agree that the structure of the requirements (left side of the V) should follow that tree requirement. However, what about if we would like to model a hierarchical element to the verification / tests on the right side of the V _at the same time_? E.g., the fact that a SW-VER may have both the SW-RS and SYS-VER as parents? It seems like Doorstop does not allow to establish this structure currently.

**Comment by sevendays** [comment]
Let me make a distinction between **verification** and **validation**:
- **verification specification** is child of previous requirements: you need to (specify how to) verify that the new level is a correct implementation of the parent level (e.g. you're designing the right thing)
- **validation specification** is child of same-level requirements: you need to (specify how to) test the product to see if its requirements are satisfied (e.g. you're building things in the right way)

That said, my structure would look like this (HW left out for clarity):

![](http://www.plantuml.com/plantuml/svg/bPF1Zjf038RlUGgBIo5TqFODLggMyuf4g0SkmuoJJ3MPoSuuKDw-JaW82Tbjzu2mFFRlpx7z4aXugaiJPTeWlEqI404MSVpctM-K10epRJLfP-FuO0-M2XsWwXv38Oo18u8HHpIe88wLSpwGg-BuQJPOkh8o5pXhAZZwzUShx7uyFyVnPhDPN1csDuPjRMLJMHZm-5vhZoLQ2X0gb3hJKjnZ8N3LjS4J6iYwn05G2ZFBXz0xfacChMg9dA9jvdmfshTV8zBz3Ea9_Otj8z9ijh61zkmr4LhG5goWsdES4JQlHOu7Yq8M_rJWFyTLKG5JJ4eg33fdabdt897LXWSbdRKeMMXeE-q6bA8ih9QSkWoNG5YkO8Vl2RVS1-u_3DoZlmjCEiN4PNGM7YSwoNxoFCbE9y_Q4kQ-_UO7Awm2ma3QvkDnz91zTalsjTnj-hW0zmeySVvDLAOBaV8JSEqg03au2QEL86o6DbhNOJ-NtuF8a2vzBFjMYoXgRdUzRixdPkwJQDkPQH95rh6krtb1uBBsjIaYMFMaMOvzl7Wj9lj7YnoDk44v3C1gO15L6nn4YTVZY-vFzGaqVGimAHFX7vJrR25sYAji-4mNbiAg_aJ38wx7odbQRIAqgkdweVVqyzwNTqC4BhMAXcyzdk5_i0QaPh97P60Cx1gekkdmHTrCdahQcGpPstmXDvCZaixaY1SMgqlp5m00)
